pkg/controllers/consolidation/controller.go

```diff
@@ -0,0 +1,552 @@
+	if !c.cloudProvider.ShouldConsolidate(node.Node, newNodes[0].Requirements, newNodes[0].InstanceTypeOptions) {
+		return consolidationAction{result: consolidateResultNotPossible}, nil
+	}
+
+	savings := nodePrice
+	// savings is reduced by the price of the new node
+	savings -= newNodes[0].InstanceTypeOptions[0].Price()
+
+	return consolidationAction{
+		oldNodes:        []*v1.Node{node.Node},
+		disruptionCost:  disruptionCost(ctx, node.pods),
+		savings:         savings,
+		result:          consolidateResultReplace,
+		replacementNode: newNodes[0],
+	}, nil
+}
+
+func (c *Controller) hasPendingPods(ctx context.Context) bool {
+	var podList v1.PodList
+	if err := c.kubeClient.List(ctx, &podList, client.MatchingFields{"spec.nodeName": ""}); err != nil {
+		// failed to list pods, assume there must be pending as it's harmless and just ensures we wait longer
+		return true
+	}
+	for i := range podList.Items {
+		if pod.IsProvisionable(&podList.Items[i]) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *Controller) replicaSetsReady(ctx context.Context) bool {
+	var rsList appsv1.ReplicaSetList
+	if err := c.kubeClient.List(ctx, &rsList); err != nil {
+		// failed to list, assume there must be one non-ready as it's harmless and just ensures we wait longer
+		return true
+	}
+	for _, rs := range rsList.Items {
+		desired := aws.Int32Value(rs.Spec.Replicas)
+		if rs.Spec.Replicas == nil {
+			// unspecified defaults to 1
+			desired = 1
+		}
+		if rs.Status.ReadyReplicas < desired {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) replicationControllersReady(ctx context.Context) bool {
+	var rsList v1.ReplicationControllerList
+	if err := c.kubeClient.List(ctx, &rsList); err != nil {
+		// failed to list, assume there must be one non-ready as it's harmless and just ensures we wait longer
+		return true
+	}
+	for _, rs := range rsList.Items {
+		desired := aws.Int32Value(rs.Spec.Replicas)
+		if rs.Spec.Replicas == nil {
+			// unspecified defaults to 1
+			desired = 1
+		}
+		if rs.Status.ReadyReplicas < desired {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) statefulSetsReady(ctx context.Context) bool {
+	var sslist appsv1.StatefulSetList
+	if err := c.kubeClient.List(ctx, &sslist); err != nil {
+		// failed to list, assume there must be one non-ready as it's harmless and just ensures we wait longer
+		return true
+	}
+	for _, rs := range sslist.Items {
+		desired := aws.Int32Value(rs.Spec.Replicas)
+		if rs.Spec.Replicas == nil {
+			// unspecified defaults to 1
+			desired = 1
+		}
+		if rs.Status.ReadyReplicas < desired {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Controller) stabilizationWindow(ctx context.Context) time.Duration {
+	// no pending pods, and all replica sets/replication controllers are reporting ready so quickly consider another consolidation
+	if !c.hasPendingPods(ctx) && c.replicaSetsReady(ctx) &&
+		c.replicationControllersReady(ctx) && c.statefulSetsReady(ctx) {
```

Comment:
What is the motivation for this check and to have a 5 min. delay if it doesn't pass?

I worry that a single broken deployment can block/delay consolidation, which could be quite expensive if you have a huge scale up and later a big scale down leading to many nodes needing to be consolidated.

In our environment we have many teams deploying to a single cluster and some teams may have a broken deployment which is not getting fixed because there is no impact from their perspective, ofc. this could be considered a problem, but that it can block/delay the consolidation is probably a side effect that is not obvious to those teams/users.

The check for pending pods seems reasonable as it doesn't consider pods that are unschedulable to my understanding. :+1:

Reply:
I imagine this is also a problem if you have rolling updates of deployments. We have many teams deploying all the time during the day, so if a rolling update would lead to a 5 min. delay that wouldn't be ideal.

Reply:
We are trying to be conservative during consolidation so that we don't disrupt the cluster too much but still drive towards a lower cost.  This looks at pending pods, deployments, etc. in an attempt to speed up consolidation if the overall cluster state looks good with respect to the workloads running on it.

Even with the 5 minute delay, you're still looking at consolidating ~12 nodes per hour or ~288 nodes per day.  We also consolidate empty nodes in a single batch which should help.

There may be tunables in the future, but in the initial version we are trying to provide a simple interface.

Reply:
We currently can scale 100s of nodes within minutes and expect them to be gone/consolidated soon after. Waiting hours/days for complete consolidation seems like a lot :)

This hardcoded 5 min. also have the effect that if nodes are empty within this time they have to wait the full window before being consolidated from my understanding. With `TTLSecondsAfterEmpty` you could reduce this to seconds, but that is no longer possible with consolidation enabled.

I can also fully understand starting with a simple interface, just trying to give a bit of feedback from a user perspective as we're eagerly awaiting this consolidation feature for likely replacing the cluster-autoscaler. For context we run 200+ Kubernetes clusters (test+prod) with up to 1000 nodes per cluster (scaling a lot during the day), so reducing the time nodes are running empty has a big compounding cost efficiency effect. :)


I'm really impressed with what karpenter has become by now with this feature, and the potential that it has compared to the cluster-autoscaler. Great work! :+1:

Reply:
There's definitely some trade-off in just about everything. I hope to get another snapshot out tomorrow (they're linked on the issue at #1091).  If you've got a dev/test cluster available that isn't critical, we would welcome some feedback on how it performs for you.